bridges/snowbridge/primitives/outbound-queue/src/v2/converter/convert.rs

```diff
@@ -0,0 +1,270 @@
+	}
+
+	fn network_matches(&self, network: &Option<NetworkId>) -> bool {
+		if let Some(network) = network {
+			*network == self.ethereum_network
+		} else {
+			true
+		}
+	}
+
+	/// Extract the fee asset item from PayFees(V5)
+	fn extract_remote_fee(&mut self) -> Result<u128, XcmConverterError> {
+		use XcmConverterError::*;
+		let _ = match_expression!(self.next()?, WithdrawAsset(fee), fee)
+			.ok_or(WithdrawAssetExpected)?;
+		let fee_asset =
+			match_expression!(self.next()?, PayFees { asset: fee }, fee).ok_or(InvalidFeeAsset)?;
+		let (fee_asset_id, fee_amount) = match fee_asset {
+			Asset { id: asset_id, fun: Fungible(amount) } => Some((asset_id, *amount)),
+			_ => None,
+		}
+		.ok_or(AssetResolutionFailed)?;
+		// Check the fee asset is Ether
+		ensure!(fee_asset_id.0 == Here.into(), InvalidFeeAsset);
+		Ok(fee_amount)
+	}
+
+	/// Convert the xcm into the Message which will be executed
+	/// on Ethereum Gateway contract, we expect an input of the form:
+	/// # WithdrawAsset(WETH)
+	/// # PayFees(WETH)
+	/// # ReserveAssetDeposited(PNA) | WithdrawAsset(ENA)
+	/// # AliasOrigin(Origin)
+	/// # DepositAsset(PNA|ENA)
+	/// # Transact() ---Optional
+	/// # SetTopic
+	fn to_ethereum_message(&mut self) -> Result<Message, XcmConverterError> {
+		use XcmConverterError::*;
+
+		// Get fee amount
+		let fee_amount = self.extract_remote_fee()?;
+
+		// Get ENA reserve asset from WithdrawAsset.
+		let mut enas =
+			match_expression!(self.peek(), Ok(WithdrawAsset(reserve_assets)), reserve_assets);
+		if enas.is_some() {
+			let _ = self.next();
+		}
+
+		// Get PNA reserve asset from ReserveAssetDeposited
+		let pnas = match_expression!(
+			self.peek(),
+			Ok(ReserveAssetDeposited(reserve_assets)),
+			reserve_assets
+		);
+		if pnas.is_some() {
+			let _ = self.next();
+		}
+
+		// Try to get ENA again if it is after PNA
+		if enas.is_none() {
+			enas =
+				match_expression!(self.peek(), Ok(WithdrawAsset(reserve_assets)), reserve_assets);
+			if enas.is_some() {
+				let _ = self.next();
+			}
+		}
+		// Check AliasOrigin.
+		let origin_location = match_expression!(self.next()?, AliasOrigin(origin), origin)
+			.ok_or(AliasOriginExpected)?;
+		let origin = AgentIdOf::convert_location(origin_location).ok_or(InvalidOrigin)?;
+
+		let (deposit_assets, beneficiary) = match_expression!(
+			self.next()?,
+			DepositAsset { assets, beneficiary },
+			(assets, beneficiary)
+		)
+		.ok_or(DepositAssetExpected)?;
+
+		// assert that the beneficiary is AccountKey20.
+		let recipient = match_expression!(
+			beneficiary.unpack(),
+			(0, [AccountKey20 { network, key }])
+				if self.network_matches(network),
+			H160(*key)
+		)
+		.ok_or(BeneficiaryResolutionFailed)?;
+
+		// Make sure there are reserved assets.
+		if enas.is_none() && pnas.is_none() {
+			return Err(NoReserveAssets)
+		}
+
+		let mut commands: Vec<Command> = Vec::new();
+
+		// ENA transfer commands
+		if let Some(enas) = enas {
```

Comment:
@yrong please factor this ENA block into a separate function, for ease of reviewing.

Reply:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/65be4090aa811befe67e4d8ec3c384505d907787